File: src/algorithms/dijkstra.js
export const dijkstra = (graph, startNodeId, endNodeId) => {
  const distances = {};
  const previous = {};
  const unvisited = new Set();
  const visitedOrder = [];

  // Initialize
  Object.keys(graph).forEach(nodeId => {
    distances[nodeId] = Infinity;
    previous[nodeId] = null;
    unvisited.add(nodeId);
  });
  distances[startNodeId] = 0;
  while (unvisited.size > 0) {
    // Find node with smallest distance
    let currentNodeId = null;
    let smallestDistance = Infinity;
    for (const nodeId of unvisited) {
      if (distances[nodeId] < smallestDistance) {
        smallestDistance = distances[nodeId];
        currentNodeId = nodeId;
      }
    }
    if (currentNodeId === null || distances[currentNodeId] === Infinity) {
      break; // No reachable nodes left or target unreachable
    }
    if (currentNodeId === endNodeId) {
      visitedOrder.push(currentNodeId);
      break; // Reached target
    }
    unvisited.delete(currentNodeId);
    visitedOrder.push(currentNodeId);

    // Neighbors
    const neighbors = graph[currentNodeId].neighbors;
    for (const neighborId in neighbors) {
      if (unvisited.has(neighborId)) {
        const alt = distances[currentNodeId] + neighbors[neighborId];
        if (alt < distances[neighborId]) {
          distances[neighborId] = alt;
          previous[neighborId] = currentNodeId;
        }
      }
    }
  }

  // Reconstruct path
  const path = [];
  let u = endNodeId;
  if (previous[u] !== null || u === startNodeId) {
    while (u !== null) {
      path.unshift(u);
      u = previous[u];
    }
  }
  return {
    visitedOrder,
    path,
    distance: distances[endNodeId]
  };
};

File: java_algorithms/Dijkstra.java
package java_algorithms;

import java.util.*;

public class Dijkstra {

    public static Result findPath(Map<String, Node> graph, String startNodeId, String endNodeId) {
        Map<String, Double> distances = new HashMap<>();
        Map<String, String> previous = new HashMap<>();
        Set<String> unvisited = new HashSet<>();
        List<String> visitedOrder = new ArrayList<>();

        // Initialize
        for (String nodeId : graph.keySet()) {
            distances.put(nodeId, Double.POSITIVE_INFINITY);
            previous.put(nodeId, null);
            unvisited.add(nodeId);
        }
        distances.put(startNodeId, 0.0);

        while (!unvisited.isEmpty()) {
            // Find node with smallest distance
            String currentNodeId = null;
            double smallestDistance = Double.POSITIVE_INFINITY;

            for (String nodeId : unvisited) {
                double dist = distances.get(nodeId);
                if (dist < smallestDistance) {
                    smallestDistance = dist;
                    currentNodeId = nodeId;
                }
            }

            if (currentNodeId == null || distances.get(currentNodeId) == Double.POSITIVE_INFINITY) {
                break; // No reachable nodes left or target unreachable
            }

            if (currentNodeId.equals(endNodeId)) {
                visitedOrder.add(currentNodeId);
                break; // Reached target
            }

            unvisited.remove(currentNodeId);
            visitedOrder.add(currentNodeId);

            // Neighbors
            Map<String, Double> neighbors = graph.get(currentNodeId).neighbors;
            for (Map.Entry<String, Double> entry : neighbors.entrySet()) {
                String neighborId = entry.getKey();
                double weight = entry.getValue();

                if (unvisited.contains(neighborId)) {
                    double alt = distances.get(currentNodeId) + weight;
                    if (alt < distances.get(neighborId)) {
                        distances.put(neighborId, alt);
                        previous.put(neighborId, currentNodeId);
                    }
                }
            }
        }

        // Reconstruct path
        List<String> path = new ArrayList<>();
        String u = endNodeId;
        if (previous.get(u) != null || u.equals(startNodeId)) {
            while (u != null) {
                path.add(0, u);
                u = previous.get(u);
            }
        }

        return new Result(visitedOrder, path, distances.get(endNodeId));
    }
}
